text/0034-ternary-hash.md

```diff
@@ -0,0 +1,498 @@
+    fn reset(&mut self);
+
+    /// Squeeze the sponge into a buffer.
+    fn squeeze_into(&mut self, buf: &mut Trits) -> Result<(), Self::Error>;
+
+    /// Convenience function using `Sponge::squeeze_into` to return an owned version of the hash.
+    fn squeeze(&mut self) -> Result<TritBuf, Self::Error> {
+        let mut output = TritBuf::zeros(HASH_LENGTH);
+        self.squeeze_into(&mut output)?;
+        Ok(output)
+    }
+
+    /// Convenience function to absorb `input`, squeeze the sponge into a buffer `buf`, and reset the sponge in one go.
+    fn digest_into(&mut self, input: &Trits, buf: &mut Trits) -> Result<(), Self::Error> {
+        self.absorb(input)?;
+        self.squeeze_into(buf)?;
+        self.reset();
+        Ok(())
+    }
+
+    /// Convenience function to absorb `input`, squeeze the sponge, and reset the sponge in one go.
+    /// Returns an owned version of the hash.
+    fn digest(&mut self, input: &Trits) -> Result<TritBuf, Self::Error> {
+        self.absorb(input)?;
+        let output = self.squeeze()?;
+        self.reset();
+        Ok(output)
+    }
+}
+```
+
+Following the sponge metaphor, an input provided by the user is `absorb`ed, and an output will be `squeeze`d from the
+data structure. `digest` is a convenience method calling `absorb` and `squeeze` in one go. The `*_into` versions of
+these methods are for passing a buffer, into which the calculated hashes are written. The internal state will not be
+cleared unless `reset` is called.
+
+### Design of `CurlP`
+
+`CurlP` is designed as a hash function, that acts on a `T1B1` binary-encoded ternary buffer, with a hash length of `243`
+trits and an inner state of `729`:
+
+```rust
+const STATE_LENGTH: usize = HASH_LENGTH * 3;
+```
+
+In addition, a lookup table is used as part of the absorption step:
+
+||-1|0|1|
+|-|-|-|-|
+|**-1**|1|1|-1|
+|**0**|0|-1|1|
+|**1**|-1|0|0|
+
+Given two balanced trits `t` and `t'`, the following table can easily be accessed by shifting them to unbalanced trits:
+```rust
+const TRUTH_TABLE: [[i8; 3]; 3] = [[1, 0, -1], [1, -1, 0], [-1, 1, 0]];
+```
+
+The way `CurlP` is defined, it can not actually fail, because the input or outputs can be of arbitrary size; hence,
+the associated type `Error = Infallible`.
+
+`CurlP` has two common variants depending on the number of rounds of hashing to apply before a hash is squeezed.
+```rust
+#[derive(Copy, Clone)]
+enum CurlPRounds {
+    Rounds27 = 27,
+    Rounds81 = 81,
+}
+```
+
+Type definition:
+
+```rust
+struct CurlP {
+    /// The number of rounds of hashing to apply before a hash is squeezed.
+    rounds: CurlPRounds,
+
+    /// The internal state.
+    state: TritBuf,
+
+    /// Workspace for performing transformations.
+    work_state: TritBuf,
+}
+```
+
+```rust
+impl Sponge for CurlP {
+    type Error = Infallible;
+    ...
+}
+```
+
+In addition, the two wrapper types for the very common `CurlP` variants with `27` and `81` rounds:
```

Comment:
Is it possible to explain a little more about the justification for these wrapper types?